web-common/src/features/dashboards/state-managers/cascading-explore-state-merge.ts

```diff
@@ -0,0 +1,64 @@
+import type { MetricsExplorerEntity } from "@rilldata/web-common/features/dashboards/stores/metrics-explorer-entity";
+
+const OneLevelDeepShallowMergeKeys = new Set<keyof MetricsExplorerEntity>([
+  "selectedTimeRange",
+  "selectedComparisonTimeRange",
+  "tdd",
+  "pivot",
+]);
+
+export function cascadingExploreStateMerge(
+  exploreStatesInOrder: Partial<MetricsExplorerEntity>[],
+) {
+  const finalExploreState: Partial<MetricsExplorerEntity> = {};
+
+  const shallowKeyProcessed = new Set<string>();
+  // Merge all keys not part of OneLevelDeepShallowMergeKeys. This allows for future keys to be merged without changes.
+  exploreStatesInOrder.forEach((state) => {
+    Object.keys(state).forEach((key: keyof MetricsExplorerEntity) => {
+      // Since the states are in order a key found 1st should only be merged once.
+      // So ignore keys we have already seen
+      const isKeyAlreadyProcessed = shallowKeyProcessed.has(key);
+
+      // Ignore one level deep merges, they are merged separately
+      const isKeyForDeepMerge = OneLevelDeepShallowMergeKeys.has(key);
```

Comment:
`isKeyForDeepMerge`... the keys will be _shallow_ merged, right? It's confusing seeing the words "DeepShallow" right next to each other.